sql/src/main/java/io/crate/operation/reference/sys/node/ArrayTypeRowContextCollectorExpression.java

```diff
@@ -0,0 +1,53 @@
+/*
+ * Licensed to Crate under one or more contributor license agreements.
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.  Crate licenses this file
+ * to you under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.operation.reference.sys.node;
+
+import com.google.common.base.Function;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Lists;
+import io.crate.metadata.RowContextCollectorExpression;
+
+import javax.annotation.Nullable;
+import java.util.List;
+
+
+abstract class ArrayTypeRowContextCollectorExpression<RowType, IterType, ReturnType>
+    extends RowContextCollectorExpression<RowType, ReturnType[]> {
+
+    protected abstract List<IterType> items();
+    protected abstract ReturnType valueForItem(IterType input);
+
+    @Override
+    public ReturnType[] value() {
+        List<IterType> items = items();
+        List<ReturnType> values = Lists.transform(items, new Function<IterType, ReturnType>() {
```

Comment:
I think the function could be a attribute to be re-used in case value() is called multiple times